internal/scraper/scraper_test.go

```diff
@@ -0,0 +1,457 @@
+package scraper
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+	bbeconfig "github.com/prometheus/blackbox_exporter/config"
+	"github.com/prometheus/blackbox_exporter/prober"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	promconfig "github.com/prometheus/common/config"
+	"github.com/prometheus/common/expfmt"
+	"github.com/stretchr/testify/require"
+)
+
+var updateGolden = flag.Bool("update-golden", false, "update golden files")
+
+func TestProber(t *testing.T) {
+	testcases := map[string]struct {
+		setup  func(t *testing.T) (string, func())
+		config bbeconfig.Module
+		prober prober.ProbeFn
+	}{
+		"ping": {
+			prober: prober.ProbeICMP,
+			setup: func(*testing.T) (string, func()) {
+				return "127.0.0.1", func() {}
+			},
+			config: bbeconfig.Module{
+				Prober:  "icmp",
+				Timeout: 2 * time.Second,
+				ICMP: bbeconfig.ICMPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+				},
+			},
+		},
+
+		"http": {
+			prober: prober.ProbeHTTP,
+			setup: func(*testing.T) (string, func()) {
+				httpSrv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					w.WriteHeader(http.StatusOK)
+				}))
+				// httpSrv.EnableHTTP2 = true
+				httpSrv.Start()
+				return httpSrv.URL, httpSrv.Close
+			},
+			config: bbeconfig.Module{
+				Prober:  "http",
+				Timeout: 2 * time.Second,
+				HTTP: bbeconfig.HTTPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+					HTTPClientConfig:   promconfig.HTTPClientConfig{},
+				},
+			},
+		},
+
+		"http_ssl": {
+			prober: prober.ProbeHTTP,
+			setup: func(*testing.T) (string, func()) {
+				httpSrv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					w.WriteHeader(http.StatusOK)
+				}))
+				// httpSrv.EnableHTTP2 = true
+				httpSrv.StartTLS()
+				return httpSrv.URL, httpSrv.Close
+			},
+			config: bbeconfig.Module{
+				Prober:  "http",
+				Timeout: 2 * time.Second,
+				HTTP: bbeconfig.HTTPProbe{
+					IPProtocol:         "ip4",
+					IPProtocolFallback: false,
+					HTTPClientConfig: promconfig.HTTPClientConfig{
+						TLSConfig: promconfig.TLSConfig{
+							InsecureSkipVerify: true,
+						},
+					},
+				},
+			},
+		},
+
+		"dns": {
+			prober: prober.ProbeDNS,
+			setup: func(*testing.T) (string, func()) {
```

Comment:
nit: can we move all this logic into a function, and call that instead of doing it in a anon func?

it would make it easy to read

Reply:
OK, done.